File: Benchmarks/OOP/OOP.php
<?php


class Vector3 {
    public float $x;
    public float $y;
    public float $z;

    public function __construct(float $x, float $y, float $z) {
        $this->x = $x;
        $this->y = $y;
        $this->z = $z;
    }

    public function magnitude_squared(): float {
        return $this->x * $this->x + $this->y * $this->y + $this->z * $this->z;
    }

    public function add(Vector3 $other): Vector3 {
        return new Vector3($this->x + $other->x, $this->y + $other->y, $this->z + $other->z);
    }
}


trait Loggable {
    public int $log_count = 0;
    public function log_message(string $message): void {
        $this->log_count++;
    }
    public function get_log_count(): int {
        return $this->log_count;
    }
}


abstract class Shape {
    use Loggable;
    public float $x;
    public float $y;

    public function __construct(float $x, float $y) {
        $this->x = $x;
        $this->y = $y;
    }

    abstract public function area(): float;
    abstract public function perimeter(): float;
}


class Circle extends Shape {
    public float $radius;

    public function __construct(float $x, float $y, float $radius) {
        parent::__construct($x, $y);
        $this->radius = $radius;
    }

    public function area(): float {
        return 3.141592653589793 * $this->radius * $this->radius;
    }

    public function perimeter(): float {
        return 2.0 * 3.141592653589793 * $this->radius;
    }
}


class Rectangle extends Shape {
    public float $width;
    public float $height;

    public function __construct(float $x, float $y, float $width, float $height) {
        parent::__construct($x, $y);
        $this->width = $width;
        $this->height = $height;
    }

    public function area(): float {
        return $this->width * $this->height;
    }

    public function perimeter(): float {
        return 2.0 * ($this->width + $this->height);
    }
}


class FileHandle {
    public string $path;
    public bool $is_open;

    public function __construct(string $path) {
        $this->path = $path;
        $this->is_open = true;
    }

    public function read(): string {
        return $this->is_open ? "Content" : "Closed";
    }

    public function close(): void {
        $this->is_open = false;
    }

    public function __destruct() {
        $this->close();
    }
}


echo "Running Struct Benchmark...\n";
$v1 = new Vector3(1.0, 2.0, 3.0);
$v2 = new Vector3(4.0, 5.0, 6.0);
$sum_struct = 0.0;
$i = 0;
while ($i < 10000000) {
    $res = $v1->add($v2);
    $sum_struct += $res->magnitude_squared();
    $i++;
}
echo "Struct: " . $sum_struct . "\n";

$sum_area = 0.0;
$j = 0;
while ($j < 5000000) {
    if ($j % 2 == 0) {
        $shape = new Circle(0.0, 0.0, 5.0);
    } else {
        $shape = new Rectangle(0.0, 0.0, 10.0, 20.0);
    }
    $sum_area += $shape->area();
    $j++;
}
echo "Polymorphism: " . $sum_area . "\n";

$k = 0;
while ($k < 3000000) {
    $file = new FileHandle("/tmp/test.txt");
    $file->read();
    $k++;
}
echo "RAII Completed.\n";

File: Benchmarks/OOP/OOP.lua
local Vector3 = {}
Vector3.__index = Vector3

function Vector3.new(x, y, z)
    local self = setmetatable({}, Vector3)
    self.x = x
    self.y = y
    self.z = z
    return self
end

function Vector3:magnitude_squared()
    return self.x * self.x + self.y * self.y + self.z * self.z
end

function Vector3:add(other)
    return Vector3.new(self.x + other.x, self.y + other.y, self.z + other.z)
end


local Loggable = {}
Loggable.__index = Loggable

function Loggable.init(self)
    self.log_count = 0
end

function Loggable:log_message(message)
    self.log_count = self.log_count + 1
end

function Loggable:get_log_count()
    return self.log_count
end


local Shape = {}
Shape.__index = Shape
setmetatable(Shape, { __index = Loggable })

function Shape.init(self, x, y)
    Loggable.init(self)
    self.x = x
    self.y = y
end


local Circle = {}
Circle.__index = Circle
setmetatable(Circle, { __index = Shape })

function Circle.new(x, y, radius)
    local self = setmetatable({}, Circle)
    Shape.init(self, x, y)
    self.radius = radius
    return self
end

function Circle:area()
    return 3.141592653589793 * self.radius * self.radius
end

function Circle:perimeter()
    return 2.0 * 3.141592653589793 * self.radius
end


local Rectangle = {}
Rectangle.__index = Rectangle
setmetatable(Rectangle, { __index = Shape })

function Rectangle.new(x, y, width, height)
    local self = setmetatable({}, Rectangle)
    Shape.init(self, x, y)
    self.width = width
    self.height = height
    return self
end

function Rectangle:area()
    return self.width * self.height
end

function Rectangle:perimeter()
    return 2.0 * (self.width + self.height)
end


local FileHandle = {}
FileHandle.__index = FileHandle

function FileHandle.new(path)
    local self = setmetatable({}, FileHandle)
    self.path = path
    self.is_open = true
    return self
end

function FileHandle:read()
    if self.is_open then
        return "Content"
    else
        return "Closed"
    end
end

function FileHandle:finalize()
    self.is_open = false
end


local v1 = Vector3.new(1.0, 2.0, 3.0)
local v2 = Vector3.new(4.0, 5.0, 6.0)
local sum_struct = 0.0
local i = 0
while i < 10000000 do
    local res = v1:add(v2)
    sum_struct = sum_struct + res:magnitude_squared()
    i = i + 1
end
print("Struct: " .. sum_struct)


local sum_area = 0.0
local j = 0
while j < 5000000 do
    local shape
    if j % 2 == 0 then
        shape = Circle.new(0.0, 0.0, 5.0)
    else
        shape = Rectangle.new(0.0, 0.0, 10.0, 20.0)
    end
    sum_area = sum_area + shape:area()
    j = j + 1
end
print("Polymorphism: " .. sum_area)


local k = 0
while k < 3000000 do
    local file = FileHandle.new("/tmp/test.txt")
    file:read()
    file:finalize()
    k = k + 1
end
print("RAII Benchmark Completed.")
